src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -121,6 +135,150 @@ public CourthouseEntity addCourtHouse(Courthouse courthouse) {
+        final SecurityRoleEntity securityRoleRequester = getSecurityRoleByRoleName("REQUESTER");
+        final SecurityRoleEntity securityRoleApprover = getSecurityRoleByRoleName("APPROVER");
+
+        //check duplicate names
+        checkCourthouseNameIsUnique(courthousePost);
+        checkCourthouseDisplayNameIsUnique(courthousePost);
+
+        List<Integer> securityGroupIds = (courthousePost.getSecurityGroupIds() == null) ? new ArrayList<>() : courthousePost.getSecurityGroupIds();
+        courthousePost.securityGroupIds(securityGroupIds);
+
+        //for each security_group get role_ids
+        List<Integer> securityRoleIds = new ArrayList<>();
+        for (Integer secGrpIds : securityGroupIds) {
+            Optional<SecurityGroupEntity> securityGroupEntity = securityGroupRepository.findById(secGrpIds);
+            securityGroupEntity.ifPresent(groupEntity -> securityRoleIds.add(groupEntity.getSecurityRoleEntity().getId()));
+        }
+
+        //ensure if security roles exist, TRANSCRIBER role also exists
+        if (securityRoleTranscriber != null && (!securityRoleIds.isEmpty() && !securityRoleIds.contains(securityRoleTranscriber.getId()))) {
+                throw new DartsApiException(AuthorisationError.BAD_REQUEST_ONLY_TRANSCIBER_ROLES_MAY_BE_ASSIGNED);
+
+        }
+
+        CourthouseEntity mappedEntity = getMappedEntityWithRegion(courthousePost);
+
+        Set<SecurityGroupEntity> securityGroupEntities = getSecurityGroupEntities(courthousePost, mappedEntity,
+                                                                                  securityRoleIds, securityRoleRequester,
+                                                                                  securityRoleApprover);
+        mappedEntity.setSecurityGroups(securityGroupEntities);
+
+        courthouseRepository.saveAndFlush(mappedEntity);
+
+        return getExtendedCourthousePost(courthousePost, mappedEntity);
+    }
+
+    private ExtendedCourthousePost getExtendedCourthousePost(CourthousePost courthousePost, CourthouseEntity mappedEntity) {
+        ExtendedCourthousePost extendedCourthousePost = courthousePostMapper.mapFromEntityToExtendedCourthousePost(courthousePost);
+
+        extendedCourthousePost.setId(mappedEntity.getId());
+        extendedCourthousePost.setCreatedDateTime(mappedEntity.getCreatedDateTime());
+        extendedCourthousePost.setLastModifiedDateTime(mappedEntity.getLastModifiedDateTime());
+        extendedCourthousePost.securityGroupIds(securityGroupIdMapper.mapSecurityGroupEntitiesToIds(mappedEntity.getSecurityGroups()));
+        extendedCourthousePost.setRegionId(mappedEntity.getRegion() == null ? null : mappedEntity.getRegion().getId());
+
+        return extendedCourthousePost;
+    }
+
+    private CourthouseEntity getMappedEntityWithRegion(CourthousePost courthousePost) {
+        CourthouseEntity mappedEntity = courthousePostMapper.mapToEntity(courthousePost);
+
+        //ensure region exists in dB
```

Comment:
Can this validation check be done earlier, e.g. in the proposed `validate()` method?

Reply:
I do not feel it can. Please also see comments below.

Reply:
I have addressed this